### PR TITLE
TF version of the trainer

### PR DESCRIPTION
**Pull request by jplu**
Tensorflow version of the PyTorch Trainer found [here](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py). It follow as much as possible the same API, usage and internal method names.

An example can be found [here](https://github.com/jplu/transformers/blob/tf-trainer/examples/ner/run_tf_ner.py). The code contains some hacks mostly for handling datasets see for example the [`TFDataset`](https://github.com/jplu/transformers/blob/tf-trainer/examples/ner/utils_ner.py#L141) that tries to imitate the `Dataset` type from PyTorch.

### CODE REVIEW

File: src/transformers/trainer_tf.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+
+
+class TFTrainer:
+    model: TFPreTrainedModel
+    args: TFTrainingArguments
+    # something similar to a PT Dataset.
+    # This is just temporary before to have
+    # a framework-agnostic approach for datasets.
+    train_dataset: Optional[TFDataset]
+    eval_dataset: Optional[TFDataset]
```
Comment (julien-c) [inline src/transformers/trainer_tf.py:37]: Shouldn't we pass `tf.data.Dataset` directly?

And then we don't need the class above?

File: src/transformers/training_tf_args.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import logging
```
Comment (julien-c) [inline src/transformers/training_tf_args.py:1]: change the name of this file to `training_args_tf.py`

File: src/transformers/trainer_tf.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        logger.info("  Num Epochs = %d", self.args.num_train_epochs)
+        logger.info("  Total optimization steps = %d", self.train_steps)
+
+        for epoch in range(1, int(epochs + 1)):
+            for training_loss in self._training_steps():
+                step = iterations.numpy()
+                training_loss = tf.reduce_mean(training_loss)
+
+                with self.writer.as_default():
+                    tf.summary.scalar("loss", training_loss, step=step)
```
Comment (julien-c) [inline src/transformers/trainer_tf.py:269]: here you're writing to TensorBoard at every step, right?

File: src/transformers/trainer_tf.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+
+                if step == 1:
+                    with self.writer.as_default():
+                        tf.summary.trace_export(name="training", step=step, profiler_outdir=self.args.logging_dir)
+
+                if step % self.args.logging_steps == 0:
+                    logger.info("Epoch {} Step {} Train Loss {:.4f}".format(epoch, step, training_loss.numpy()))
+
+                    if self.args.evaluate_during_training and step % self.args.eval_steps == 0:
```
Comment (julien-c) [inline src/transformers/trainer_tf.py:278]: if `eval_steps ` is not a multiple of `logging_steps `, this will never run, no?

File: src/transformers/training_tf_args.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+@dataclass
+class TFTrainingArguments(TrainingArguments):
+    tpu: bool = field(default=False, metadata={"help": "Run the training over TPUs"})
+    optimizer_name: str = field(default="adam", metadata={"help": "Name of a Tensorflow optimizer"})
+    mode: str = field(
+        default="sequence-classification",
+        metadata={"help": 'Type of task, one of "sequence-classification", "token-classification" '},
+    )
+    loss_name: str = field(default="SparseCategoricalCrossentropy", metadata={"help": "Name of a Tensorflow loss"})
+    eval_steps: int = field(default=1000, metadata={"help": "Run an eval every X steps."})
```
Comment (julien-c) [inline src/transformers/training_tf_args.py:24]: in PyTorch we basically use `logging_steps` for this in conjunction with `--evaluate_during_training`

I think here we could do the same, no? (Also because I think there's an issue which I've mentioned in the training loop above)

**Review by julien-c — APPROVED** [review]
I love it! I guess my main question is around the type of datasets, so I'm gonna ping @thomwolf on this.

But overall I think it's a really great job!

Do you have links to Tensorboards for trainings you launched on GLUE or NER? (no worries if not, we can launch them after this)

**Comment by jplu** [comment]
Thanks a lot for these reviews :) I'm currently running the trainer over all the GLUE tasks and few  NER datasets. I still have few bugs solved that will be pushed soon, that will include your reviews.

I do have some logs in Tensorboard yes, I can share them but I suggest you do some test on your side as well to be sure it is ok.

File: src/transformers/optimization_tf.py [hunk 6]
```diff
@@ -147,16 +151,16 @@ def _resource_apply_dense(self, grad, var, apply_state=None):
...
     def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
         lr_t, kwargs = self._get_lr(var.device, var.dtype.base_dtype, apply_state)
         decay = self._decay_weights_op(var, lr_t, apply_state)
         with tf.control_dependencies([decay]):
-            return super()._resource_apply_sparse(grad, var, indices, **kwargs)
+            return super(AdamWeightDecay, self)._resource_apply_sparse(grad, var, indices, **kwargs)
 
     def get_config(self):
-        config = super().get_config()
+        config = super(AdamWeightDecay, self).get_config()
```
[the remark this answers is not in the document]
Comment (jplu) [reply src/transformers/optimization_tf.py:163]: I took this piece of code from elsewhere so I did not really touch it.

File: src/transformers/training_tf_args.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+logger = logging.getLogger(__name__)
+
+if is_tf_available():
+    import tensorflow as tf
+
+
+@dataclass
+class TFTrainingArguments(TrainingArguments):
+    tpu: bool = field(default=False, metadata={"help": "Run the training over TPUs"})
+    optimizer_name: str = field(default="adam", metadata={"help": "Name of a Tensorflow optimizer"})
```
[the remark this answers is not in the document]
Comment (jplu) [reply src/transformers/training_tf_args.py:18]: Good idea, I will put a link to the TF doc where all the names are listed.

File: src/transformers/training_tf_args.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    mode: str = field(
+        default="sequence-classification",
+        metadata={"help": 'Type of task, one of "sequence-classification", "token-classification" '},
+    )
+    loss_name: str = field(default="SparseCategoricalCrossentropy", metadata={"help": "Name of a Tensorflow loss"})
```
[the remark this answers is not in the document]
Comment (jplu) [reply src/transformers/training_tf_args.py:23]: Yep

File: src/transformers/trainer_tf.py (hunk 1 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:37]: This is temporary, waiting the new lib. I did that mostly to be able to keep the size of a dataset, as it is quite painfull to compute this in TF.

File: src/transformers/trainer_tf.py (hunk 3 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:269]: Yes, I trace only the loss at every step.

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:278]: Excactly! Good catch.

File: src/transformers/training_tf_args.py (hunk 5 above)
Comment (jplu) [reply src/transformers/training_tf_args.py:24]: It is something I'm not really fan on in the trainer, because I like to have my loss print every "X" steps but I don't want the model to be evaluated as many time as there is a log, that's why I have created a new parameter, to separate the log itself and the evaluation.

Comment (julien-c) [reply src/transformers/training_tf_args.py:24]: Ok, I kinda agree, I can change it in the Trainer too then. I wasn't a fan of this either.

Only thing is that it's breaking the compatibility with the old example scripts a little (not a huge deal)

File: src/transformers/trainer_tf.py (hunk 3 above)
Comment (julien-c) [reply src/transformers/trainer_tf.py:269]: A single step's loss (with no averaging whatsoever) is kinda random though, no? 

Shouldn't we keep the same behaviour as in Trainer?

File: src/transformers/optimization_tf.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (julien-c) [reply src/transformers/optimization_tf.py:163]: Yes but it's in the diff of optimization_tf.py :)

No big deal, we can remove after merging.

File: src/transformers/trainer_tf.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/trainer_tf.py:37]: So, we will probably launch TFTrainer before the "library-in-stealth", so it should probably be a tf.data.Dataset for now. Will be easier to understand, in my opinion. And we won't have to change anything when the new library is out.

If it's really costly to compute the dataset size, can we maybe just store the size as an attribute on the object in the scripts, and add an error message to the init of the TFTrainer if it's not set?

File: examples/run_tf_glue.py [hunk 9]
```diff
@@ -1,105 +1,240 @@
+# coding=utf-8
+""" Fine-tuning the library models for named entity recognition."""
```
Comment (LysandreJik) [inline examples/run_tf_glue.py:2]:
```suggestion
""" Fine-tuning the library models for sequence classification."""
```

File: src/transformers/trainer_tf.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/trainer_tf.py:37]: I advocate for having `tf.data.Dataset` here as well!

**Review by LysandreJik — APPROVED** [review]
Great! Nice to see how much code can be removed from the NER example.

**Comment by jplu** [comment]
@julien-c As asked here few Tensorboards:

- Sequence Classification example with [GLUE mrpc](https://tensorboard.dev/experiment/xJt1nogwRDO1RkebX6Qspg/#scalars) 
- Regression example with [GLUE sts-b](https://tensorboard.dev/experiment/szdpv9BzTVq8euJKBZpuAQ/#scalars)
- Token Classification example with [Germeval](https://tensorboard.dev/experiment/JMNV2qqbRRamq9EbPLo3Pg/#scalars)

The final commit will arrive during the weekend 😄

File: examples/run_tf_glue.py (hunk 9 above)
Comment (jplu) [reply examples/run_tf_glue.py:2]: Ahah the usual miss of copy/paste 😆 Will be updated for the next commit ^^

File: src/transformers/trainer_tf.py (hunk 1 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:37]: You are right, I will find a hack that will do the job^^

File: src/transformers/trainer_tf.py (hunk 3 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:269]: I like to keep trace when I run a model for the first time to see how it evolves and check what are the important steps and then afterwards I can better tune the model, but I don't mind to remove it. As you wish 😄

Comment (julien-c) [reply src/transformers/trainer_tf.py:269]: No I think we can keep it

File: src/transformers/training_tf_args.py (hunk 5 above)
Comment (thomwolf) [reply src/transformers/training_tf_args.py:24]: Sounds good to me as well (and I don't think we should really guarantee any backward compatibility for the CLI arguments of the example scripts).

**Comment by jplu** [comment]
@julien-c I'm now ok with this PR, the trainer has been tested over all the GLUE tasks + Germeval + CoNLL2002 and 2003. There is stil a non blocking minor bug for regression tasks (a raised exception that cannot be catched each time the graph is computed) that I do not really understand, so I will wrap it up into a small piece of code and open an issue in the TF repo. 

Do you want to do a last check?

**Review by thomwolf — APPROVED** [review]
Really nice @jplu!

File: src/transformers/training_args_tf.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        logger.info("Tensorflow: setting up strategy")
+        gpus = tf.config.list_physical_devices("GPU")
+
+        if self.no_cuda:
+            strategy = tf.distribute.OneDeviceStrategy(device="/cpu:0")
+        else:
+            try:
+                tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
+            except ValueError:
+                tpu = None
```
Comment (LysandreJik) [inline src/transformers/training_args_tf.py:50]: A welcome sight!

Comment (LysandreJik) [inline src/transformers/training_args_tf.py:50]: Could we possibly add a `--no_tpu` flag as well, similar to the `--no_cuda`? Talked about it with @julien-c a few minutes ago, we're probably going to do it this way in the PyTorch trainer

File: src/transformers/training_args_tf.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    @property
+    @tf_required
+    def strategy(self) -> "tf.distribute.Strategy":
+        return self._setup_strategy
+
+    @property
+    @tf_required
+    def n_gpu(self) -> int:
+        return self._setup_strategy.num_replicas_in_sync
```
Comment (LysandreJik) [inline src/transformers/training_args_tf.py:75]: Will this return the number of TPU cores when on a device having access to a TPU?

**Review by LysandreJik — APPROVED** [review]
Added a few comments related to the TPU, not too important for the merge, can be updated later. Feel free to merge if you want @julien-c.

Thanks a lot for the hard work @jplu, the trainer is great!

File: src/transformers/training_args_tf.py (hunk 10 above)
Comment (jplu) [reply src/transformers/training_args_tf.py:50]: Here we check if there is a TPU cluster available, if not a ValueError is raised then we put tpu to None

I'm gonna put the same comment in the code.

Comment (jplu) [reply src/transformers/training_args_tf.py:50]: Of course it is possible. So I suppose it will be in the `trainer_args.py` files?

File: src/transformers/training_args_tf.py (hunk 11 above)
Comment (jplu) [reply src/transformers/training_args_tf.py:75]: Yes

File: src/transformers/training_args_tf.py (hunk 10 above)
Comment (LysandreJik) [reply src/transformers/training_args_tf.py:50]: Yes, it's great to see TPU support here!

Comment (LysandreJik) [reply src/transformers/training_args_tf.py:50]: Indeed, in the `TrainingArguments` class! The[ PyTorch TPU trainer](https://github.com/huggingface/transformers/pull/4146/files#diff-ef17e4d152412f77ce542cec19496f81R109) currently has a `--use_tpu` flag but we'll switch it to `--no_tpu` instead later today.

File: src/transformers/training_args_tf.py (hunk 11 above)
Comment (LysandreJik) [reply src/transformers/training_args_tf.py:75]: I'm thinking it should return `0` as there's no GPU. Is this property used in places where the number of TPU cores would change behavior?

Comment (jplu) [reply src/transformers/training_args_tf.py:75]: Here it returns the number of GPUs if the GPUs are used or TPUs if they are used

Comment (LysandreJik) [reply src/transformers/training_args_tf.py:75]: Right, I guess this can be okay for the moment but down the road it would be nice to see have something clearer as the `n_gpu` property doesn't make much sense in a TPU environment.

I guess similar to the logging where it outputs `n_gpu: %s, distributed training: %s, 16-bits training: %s`, I guess we'd want to see `TPU cores: xxx ...`.

This can be in a next PR, it's mostly nit-picking!

Comment (jplu) [reply src/transformers/training_args_tf.py:75]: Ok, we can review that later indeed as I think we need a bit of discuss deeper how harmonize

Comment (julien-c) [reply src/transformers/training_args_tf.py:75]: I can rename it to num_devices everywhere if it makes sense?

**Comment by julien-c** [comment]
Awesome work @jplu! 💪💪💪

File: src/transformers/training_args_tf.py (hunk 11 above)
Comment (jplu) [reply src/transformers/training_args_tf.py:75]: Or accelerators? As you wish but indeed will have more sense to generalize here.
